### ISSUE TITLE
Are params in meta_net updated?

### ISSUE THREAD

**Issue by Cocoxili** [danieltan07/learning-to-reweight-examples#3]
It seems that the code below always get eps=[0...], l_f_meta=0 and the params in meta_net do not update in each iteration.

```
eps = to_var(torch.zeros(cost.size()))
l_f_meta = torch.sum(cost * eps)

meta_net.zero_grad()
        
# Line 6 perform a parameter update
grads = torch.autograd.grad(l_f_meta, (meta_net.params()), create_graph=True)

meta_net.update_params(hyperparameters['lr'], source_params=grads)

```

Am I getting it wrong?

**Comment by danieltan07** [comment]
@Cocoxili  sorry for the very late reply. It is updating. What we are after here is the gradient. A loss of 0 does not mean there is no gradient.

**Comment by guixianjin** [comment]
But when I print grads, it's really all zeros.

        # Line 6 perform a parameter update
        grads = torch.autograd.grad(l_f_meta, (meta_net.params()), create_graph=True)
        print(grads)
        meta_net.update_params(hyperparameters['lr'], source_params=grads)

 I think maybe  when epsilon is zeros, grads(gradient wrt. theta_t)  are zeros. When epsilon is not zeros, grads are not zeros. And what is important is  the gradient wrt. epsilon, not the gradient wrt. theta_t.
Although the current value of grads are zeros, but it is a symbol containing variable epsilon, and we need  the grads to get the gradient wrt. epsilon by chain rule.  And we don't really need to update meta_net. We only need to know how the value of epsilon influence "one-step look ahead" update.

**Comment by pengwangucla** [comment]
Same question here. I don't think the parameter is updated since the grads are all zero

**Comment by pengwangucla** [comment]
> But when I print grads, it's really all zeros.
> 
> ```
>     # Line 6 perform a parameter update
>     grads = torch.autograd.grad(l_f_meta, (meta_net.params()), create_graph=True)
>     print(grads)
>     meta_net.update_params(hyperparameters['lr'], source_params=grads)
> ```
> 
> I think maybe when epsilon is zeros, grads(gradient wrt. theta_t) are zeros. When epsilon is not zeros, grads are not zeros. And what is important is the gradient wrt. epsilon, not the gradient wrt. theta_t.
> Although the current value of grads are zeros, but it is a symbol containing variable epsilon, and we need the grads to get the gradient wrt. epsilon by chain rule. And we don't really need to update meta_net. We only need to know how the value of epsilon influence "one-step look ahead" update.

So did you reproduced the results ? I can't run out the results.

**Comment by zhouyiheng11** [comment]
I agree with @guixianjin . When calculating gradient by using chain rules, all the gradients will multiply by the value of eps, which is zero.

**Comment by danieltan07** [comment]
What @guixianjin is saying is correct. Gradient w.r.t to epsilon is what is needed and it is not necessarily 0. It may be 0 for some mini batches if there is no reason to re-weight. I didn't really check but im guessing there will be some batches that will have non zero gradients specially since it does work at least on the heavily imbalanced mnist toy problem. 

Note that, I was merely reproducing the results of the paper. Initializing epsilons to 0 is the design choice of the authors of the paper (Algorithm 1), not mine.

**Comment by zhouyiheng11** [comment]
@danieltan07  Yes, the gradient w.r.t. eps is non-zero most of the time. Thanks for sharing the implementation :)

**Comment by hongxin001** [comment]
@danieltan07 Gradient w.r.t to epsilon is non-zero, but Gradient w.r.t to params of the copied net would be zero forever. It means the update in the line 7 of algorithm 1 is zero. I think we need to check the official code for understanding the true.

**Comment by danieltan07** [comment]
@hongxin001 Yes, I agree. When epsilon is zero the gradient w.r.t to the params of the net is zero. The algorithm they provided probably has a typo or my implementation is incorrect. Maybe they meant it to be initialized really close to zero but not zero.  I guess it's best to check their official code to make sure of this.

**Comment by ghost** [comment]
@danieltan07 The following code shows how the original code initializes the weights. It can be seen that it is also initialized to 0. So the update in the line 7 of algorithm 1 actually has no effect on the parameters. But it does not affect the effectiveness of the algorithm, because starting from a certain parameter value, the weight will be updated next time, and the updated weight will once again help us find a better parameter, and so on.
```python
def reweight_autodiff(inp_a,
                      label_a,
                      inp_b,
                      label_b,
                      bsize_a,
                      bsize_b,
                      eps=0.0,
                      gate_gradients=1):
    """Reweight examples using automatic differentiation.
    :param inp_a:             [Tensor]    Inputs for the noisy pass.
    :param label_a:           [Tensor]    Labels for the noisy pass.
    :param inp_b:             [Tensor]    Inputs for the clean pass.
    :param label_b:           [Tensor]    Labels for the clean pass.
    :param bsize_a:           [int]       Batch size for the noisy pass.
    :param bsize_b:           [int]       Batch size for the clean pass.
    :param eps:               [float]     Minimum example weights, default 0.0.
    :param gate_gradients:    [int]       Tensorflow gate gradients, reduce concurrency.
    """
    ex_wts_a = tf.zeros([bsize_a], dtype=tf.float32)
    ex_wts_b = tf.ones([bsize_b], dtype=tf.float32) / float(bsize_b)
    w_dict, loss_a, logits_a = get_model(
        inp_a, label_a, ex_wts=ex_wts_a, is_training=True, reuse=True)
    var_names = w_dict.keys()
    var_list = [w_dict[kk] for kk in var_names]
    grads = tf.gradients(loss_a, var_list, gate_gradients=gate_gradients)

    var_list_new = [vv - gg for gg, vv in zip(grads, var_list)]
    w_dict_new = dict(zip(var_names, var_list_new))
    _, loss_b, logits_b = get_model(
        inp_b, label_b, ex_wts=ex_wts_b, is_training=True, reuse=True, w_dict=w_dict_new)
    grads_ex_wts = tf.gradients(loss_b, [ex_wts_a], gate_gradients=gate_gradients)[0]
    ex_weight = -grads_ex_wts
    ex_weight_plus = tf.maximum(ex_weight, eps)
    ex_weight_sum = tf.reduce_sum(ex_weight_plus)
    ex_weight_sum += tf.to_float(tf.equal(ex_weight_sum, 0.0))
    ex_weight_norm = ex_weight_plus / ex_weight_sum
    return ex_weight_norm
```
